### PR TITLE
support broker level dispatch rate limiter

### LINKED ISSUE #7720: Broker level dispatch rate limiter
**Is your feature request related to a problem? Please describe.**
Currently, the Pulsar support topic level and subscription level dispatch rate limiter. When we have more topics and consumers(with high receive queue size) attach many topics, this will lead to a sudden increase in entry read traffic.
It may bring unstable factors to the cluster(may cause broker outgoing network bandwidth with the heavy workload, may cause too many read request to bookies, etc.)

The topic level rate limiter and the subscription level rate limiter unable to limit the resources of the entire broker because of the topics and the subscriptions may keep increasing.

image

The above screenshot is happening when I test read entries with 200MB/s, but it reached 600MB/s at the moment of starting the consumer, and the message writing also been affected.

**Describe the solution you'd like**
Introduce broker level dispatch rate limiter. This will make it easier for us to do broker resource planning.

### PR DESCRIPTION
**Pull request by wangjialing218**
### Motivation

Fixes #7720

Current we support broker level publish rate limiter.
This PR will add broker level dispatch rate limiter.

### Modifications

Add broker level dispatch rate limiter for msgs and bytes.
Refact calculateToRead() to reduce duplicate code.

### Verifying this change

Add testBrokerBytesRateLimitingReceiveAllMessagesAfterThrottling()

### CODE REVIEW

**Comment by wangjialing218** [comment]
@codelipenghui @315157973 please have a look

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/SubscriptionMessageDispatchThrottlingTest.java [hunk 1]
```diff
@@ -310,6 +310,102 @@ public void testBytesRateLimitingReceiveAllMessagesAfterThrottling(SubscriptionT
         log.info("-- Exiting {} test --", methodName);
     }
 
+    /**
+     * verify broker level rate-limiting should throttle message-dispatching based on byte-rate
```
Comment (Anonymitaet) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/api/SubscriptionMessageDispatchThrottlingTest.java:314]:
```suggestion
     * Verify whether the broker level rate-limiting is throttle message-dispatching based on byte-rate or not
```

do you mean this?

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/SubscriptionMessageDispatchThrottlingTest.java [hunk 2]
```diff
@@ -310,6 +310,102 @@ public void testBytesRateLimitingReceiveAllMessagesAfterThrottling(SubscriptionT
...
+     *
+     * <pre>
+     *  1. broker level dispatch-byte-rate = 1000 bytes/sec
+     *  2. start two consumers for two topics
+     *  3. send 15 msgs to each of the two topics: each msgs with 100 byte, total 3000 byte
+     *  4. it should take up to 2 second to receive all messages of the two topic
```
Comment (Anonymitaet) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/api/SubscriptionMessageDispatchThrottlingTest.java:320]:
```suggestion
     *  1. Broker level dispatch-byte-rate is equal to 1000 bytes per second.
     *  2. Start two consumers for two topics.
     *  3. Send 15 msgs to each of the two topics. Each msgs with 100 bytes, thus 3000 bytes in total.
     *  4. It should take up to 2 seconds to receive all messages of the two topics.
```

**Comment by Anonymitaet** [comment]
Thanks for your contribution. 

The [PR template contains info about doc](https://github.com/apache/pulsar/blob/master/.github/PULL_REQUEST_TEMPLATE.md#documentation), which helps others know more about the changes. Can you provide doc info in future PR descriptions? Thanks

**Comment by wangjialing218** [comment]
@315157973 @codelipenghui @hangc0276 PTAL when you have time

**Comment by 315157973** [comment]
The current reading process is like this:
1) Determine whether the threshold is exceeded
2) Enter the pending queue and prepare to read
3) The entry is successfully read, and the threshold count is deducted

There is a time difference between 1 and 3. When the Broker just restarts, if it is read concurrently, the problem in that PR will still exists

**Comment by 315157973** [comment]
> > The current reading process is like this:
> > 
> > 1. Determine whether the threshold is exceeded
> > 2. Enter the pending queue and prepare to read
> > 3. The entry is successfully read, and the threshold count is deducted
> > 
> > There is a time difference between 1 and 3. When the Broker just restarts, if it is read concurrently, the problem in that PR will still exists
> 
> The problem you mentioned may be solved in this PR #8611

This PR will only prevent the message from being sent to the consumer for a long time later, but it cannot solve the memory surge during concurrent reading when broker restart.

**Comment by wangjialing218** [comment]
> This PR will only prevent the message from being sent to the consumer for a long time later, but it cannot solve the memory surge during concurrent reading when broker restart.

As current reading process, it's not easy to completely prevent the sudden increase entry read traffic when mutliply comsumers starting at same time. 
With PR #8611 and this PR, the increase traffic will last only one second. This could reduce the influence of write traffic in case of #7720

**Comment by wangjialing218** [comment]
@315157973 could you have a look again

**Comment by 315157973** [comment]
Some unit tests seem to fail

**Comment by wangjialing218** [comment]
@315157973 @codelipenghui could you have a check again

File: conf/broker.conf [hunk 3]
```diff
@@ -327,6 +327,14 @@ brokerPublisherThrottlingMaxMessageRate=0
 # (Disable byte rate limit with value 0)
 brokerPublisherThrottlingMaxByteRate=0
 
+# Max Rate(in 1 seconds) of Message allowed to dispatch from a broker if broker dispatch rate limiting enabled
+# (Disable message rate limit with value 0)
+brokerDispatchThrottlingMaxMessageRate=0
```
Comment (codelipenghui) [inline conf/broker.conf:332]:
```suggestion
dispatchThrottlingRateInMsg=0
```

File: conf/broker.conf [hunk 4]
```diff
@@ -327,6 +327,14 @@ brokerPublisherThrottlingMaxMessageRate=0
...
+
+# Max Rate(in 1 seconds) of Byte allowed to dispatch from a broker if broker dispatch rate limiting enabled.
+# (Disable byte rate limit with value 0)
+brokerDispatchThrottlingMaxByteRate=0
```
Comment (codelipenghui) [inline conf/broker.conf:336]:
```suggestion
dispatchThrottlingRateInByte=0
```

Comment (codelipenghui) [reply conf/broker.conf:336]: And we already have some `dispatchThrottling` related configurations such as `dispatchThrottlingRatePerTopicInByte`, it's better to put them together and follow the same name pattern

Comment (wangjialing218) [reply conf/broker.conf]: Done

**Review by codelipenghui — APPROVED** [review]

**Comment by wangjialing218** [comment]
webServicePort

**Comment by Anonymitaet** [comment]
As discussed w/ @wangjialing218, he will add the two parameters and corresponding descriptions to the [configuration - standalone](https://pulsar.apache.org/docs/en/next/reference-configuration/#standalone) page.

**Pull request merged**